### ISSUE TITLE
Add  --output-dir to the helmfile template command

### ISSUE THREAD

**Issue by lwolf** [roboll/helmfile#629]
For cases when helmfile is used only for templating, will be great to support `--output-dir` as original helm does.

```
$ helm template --help
...
      --output-dir string        writes the executed templates to files in output-dir instead of stdout
...
```
Some benefits:
* it renders yamls to corresponding files, preserving structure of the original chart
* it ignores empty files [helm/helm/issues/3979]

**Comment by mumoshu** [comment]
@lwolf Thanks for submitting this request! Yeah, this sounds like a great enhancement to helmfile.

**Comment by mumoshu** [comment]
@olivierboudet Hey! Thanks for submitting your PR #693 with this.

I wanted to discuss with you folks what the expected behavior of this should be.

Just passing `--output-dir USER_SPECIFIED_DIR` to every `helm template` command called by helmfile breaks when there are two or more releases and nested state files that uses the same chart, because it uses the name of the chart as the directory name:

```
helm template --name foo mysql-1.2.0.tgz --output-dir test1
wrote test1/mysql/templates/secrets.yaml
wrote test1/mysql/templates/tests/test-configmap.yaml
wrote test1/mysql/templates/pvc.yaml
wrote test1/mysql/templates/svc.yaml
wrote test1/mysql/templates/tests/test.yaml
wrote test1/mysql/templates/deployment.yaml
```

In other words, the question is - what if `mysql` chart is used by multiple releases, shouldn't it override the previously "templated" one?

So probably we want the chart name part in the directory name to be `<state-file-id>-<release-name>`.

We can probably use some hash of the absolute path to the state file as `state-file-id`, and append an index suffix when there are two or more file that ends up with the same `state-file-id`.

WDYT?

**Comment by mumoshu** [comment]
Also cc / @lwolf

**Comment by olivierboudet** [comment]
Yes @mumoshu I did not thought about that, just passing a user specified output dir to every `helm template` command was enough for my use case.

I agree with your proposal but I think the directory `<state-file-id>-<release-name>` should also contain the state file name so that users can easily find generated templates for each state file...

I probably miss something but if we generate a hash of the absolute path to the state file, we may consider that only one file can have the same `state-file-id`, am I wrong ?

**Comment by lwolf** [comment]
Yeah, it makes sense to have unique directory for each release inside the user speficied directory

**Comment by mumoshu** [comment]
> I probably miss something but if we generate a hash of the absolute path to the state file, we may consider that only one file can have the same state-file-id, am I wrong ?

@olivierboudet You aren't missing anything! Only one file can have the same state-file-id, yes. That's why I wrote:

> and append an index suffix when there are two or more file that ends up with the same state-file-id.

I wrote it based on my assumption that no hash function is capable of generating unique IDs(although they have low probability to collide).

WDYT?

**Comment by olivierboudet** [comment]
I pushed a new implementation with a subdirectory by release.

For a simple state file like this : 
```
releases:
  - name: my-release
    chart: stable/mysql
```

The result is : 
```
Fetching stable/mysql
wrote output/helmfile-e9b37d2f-my-release/mysql/templates/secrets.yaml
wrote output/helmfile-e9b37d2f-my-release/mysql/templates/tests/test-configmap.yaml
wrote output/helmfile-e9b37d2f-my-release/mysql/templates/pvc.yaml
wrote output/helmfile-e9b37d2f-my-release/mysql/templates/svc.yaml
wrote output/helmfile-e9b37d2f-my-release/mysql/templates/tests/test.yaml
wrote output/helmfile-e9b37d2f-my-release/mysql/templates/deployment.yaml
```

The first part is the name of the helmfile (I keep it because I have multiple helmfiles named helmfile-middlewares, helmfile-apps, helmfile-ingress and I want to see easily the output for them).
The second part is first 8 chars of the sha1 of the absolute path of the state file.
The third part is the release name.

For the moment it will not add an index if the sha1 collide with another but I will try to add it later.

PS : I am not a go developer, do not hesitate if I can improve some things

**Comment by mumoshu** [comment]
@olivierboudet Thanks a lot for your work. It looks great!

I'd appreciate it if you could all some unit test so that we don't unintentionally break it in the future. But it isn't a requisite for merging

**Comment by olivierboudet** [comment]
@mumoshu yes I will try to add some as soon as possible. As I can see, there is no unit tests for template command for now that I can enrich ?

**Comment by mumoshu** [comment]
@olivierboudet Thanks! Yeah I think there's no unit tests for `template` now.

There may be several options to start with. I'll describe one of them here.

- Create an another `TestWhatever` function copying one of existing tests in [app_test.go](https://github.com/roboll/helmfile/blob/master/pkg/app/app_test.go). It is the high-level test suite that basically calls various helmfile external/internal commands in go code and verifies the result.

- Probably you want to run `err := app.Template()` in order to test the `helmfile template` command as a whole. But it isn't that simple, unfortunately.. Just calling `app.Template()` ends up calling actual `helm` commands in your "unit" test, which isn't nice.

- If you have enough time, try changing the code a bit to allow dependency-injecting the `helmexec.Execer` implementation that is currently hard-coded at https://github.com/roboll/helmfile/blob/f61334d9bb4cf637d3a9152b1595b8d4b6462db5/pkg/app/app.go#L277. Maybe you can add `execer` as a field to [`App` struct](https://github.com/roboll/helmfile/blob/f61334d9bb4cf637d3a9152b1595b8d4b6462db5/pkg/app/app.go#L46), so that you can inject your own [mock execer](https://github.com/roboll/helmfile/blob/f61334d9bb4cf637d3a9152b1595b8d4b6462db5/pkg/state/state_test.go#L668) for testing.

**Comment by olivierboudet** [comment]
I try to add a unit test but I have some issues.

I have mocked the helmexec.Execer, so the test is running correctly. But I don't know how to do assertions.

I try to catch the logger output in debug to verify that the command (ie helm template) is correct but the path where the chart is downloaded is dynamically generated. For example : 
```
exec: helm template /tmp/707176687/myrelease1/latest/mychart1 --name myrelease1 --output-dir output/subdir/helmfile-a94eb40d-myrelease1 --kube-context dev
```
How can I verify that the command is correct ignoring the ```707176687``` part ?

My test case is as follow for the moment : 
```
func TestTemplate_SingleStateFile(t *testing.T) {
	files := map[string]string{
		"/path/to/helmfile.yaml": `
releases:
- name: myrelease1
  chart: mychart1
- name: myrelease2
  chart: mychart1
`,
	}

	var buffer bytes.Buffer
	logger := helmexec.NewLogger(&buffer, "debug")

	app := appWithFs(&App{
		glob:        filepath.Glob,
		abs:         filepath.Abs,
		KubeContext: "default",
		Env:         "default",
		Logger:      logger,
		helmExecer:  MockExecer(logger, "dev"),
	}, files)
	app.Template(configImpl{})

	expected := `Generating templates to : output/subdir/helmfile-a94eb40d-myrelease1
exec: helm template /tmp/707176687/myrelease1/latest/mychart1 --name myrelease1 --output-dir output/subdir/helmfile-a94eb40d-myrelease1 --kube-context dev
exec: helm template /tmp/707176687/myrelease1/latest/mychart1 --name myrelease1 --output-dir output/subdir/helmfile-a94eb40d-myrelease1 --kube-context dev:

exec: helm template /tmp/707176687/myrelease1/latest/mychart1 --name myrelease2 --output-dir output/subdir/helmfile-a94eb40d-myrelease2 --kube-context dev
exec: helm template /tmp/707176687/myrelease1/latest/mychart1 --name myrelease2 --output-dir output/subdir/helmfile-a94eb40d-myrelease2 --kube-context dev: 
`
	if !strings.Contains(buffer.String(), expected) {
		t.Errorf("app.Template()\nactual = %v\nexpect = %v", buffer.String(), expected)
	}
}
```

**Comment by mumoshu** [comment]
@olivierboudet Nice try!

I'd suggest enhancing `mockHelmExec` so that can capture what are "templated" like we do for other helm commands https://github.com/roboll/helmfile/blob/1df9b0e9e9cfd6230afdb3c34b9e406df58b4a37/pkg/state/state_test.go#L737

Based on that, you'll be able to assert each set of directory and flags used for calling `helm template`, like `/tmp/707176687/myrelease1/latest/mychart1` and `--name myrelease1 --output-dir output/subdir/helmfile-a94eb40d-myrelease1 --kube-context dev` respectively.

You still see `/tmp/<number>` in the directory, but as you have access to each entry it should be easier to extract the `myrelease1/latest/mychart1` part only and make assertion on that.

**Comment by bitsofinfo** [comment]
awesome!
